### PR TITLE
Lists bug

### PR DESCRIPTION
**Pull request by Schwusch**
This pull solves this problem:
```dart
final left = {
  'fictionalValue': [
    {
      'subValue': [
        'left',
      ],
    },
    {
      'subValue': [
        'same on both',
      ],
    },
  ],
};
final right = {
  'fictionalValue': [
    {
      'subValue': [
        'right',
      ],
    },
    {
      'subValue': [
        'same on both',
        'added',
      ],
    },
  ],
};

final node = JsonDiffer.fromJson(left, right).diff();
// Passed before the fix
expect(node.node['fictionalValue'].node['0'].node['subValue'].changed['0'], equals(['left', 'right']));
// Fails without the fix
expect(node.node['fictionalValue'].node['1'].node['subValue'].added['1'], equals('added'));
```

### CODE REVIEW

File: lib/src/diff_node.dart [hunk 1]
```diff
@@ -58,103 +59,112 @@ class DiffNode {
...
-      result += '$nl${gap}{$nl';
-      node.forEach((key, d) => result += '$gap${ss}$key: ${d.toString(gap: gap+'    ', pretty: pretty)}');
-      result += '$nl${gap}}';
-    }
-    return result;
-  }
+  @override
+  String toString() => _diffTexts(this, '').join('\n');
+
+  List<String> _diffTexts(DiffNode diff, String path) => [
```
Comment (srawlins) [inline lib/src/diff_node.dart:153]: Nice collection-for and spread.

File: lib/src/diff_node.dart [hunk 2]
```diff
@@ -58,103 +59,112 @@ class DiffNode {
...
   ///
   /// If a child DiffNode has nothing added, removed, changed, nor a node, then it will
   /// be deleted from the parent's [node] Map.
   void prune() {
-    List<String> keys = node.keys.toList();
-    for (int i=keys.length-1; i>=0; i--) {
-      String key = keys[i];
-      DiffNode d = node[key];
+    var keys = node.keys.toList();
+    for (var i = keys.length - 1; i >= 0; i--) {
```
Comment (srawlins) [inline lib/src/diff_node.dart:140]: Wow, this code looks like it has never been formatted with `dartfmt` before...

File: lib/src/json_differ.dart [hunk 3]
```diff
@@ -5,30 +5,50 @@ part of json_diff;
...
   /// library to encode each Map into a JSON String.
-  JsonDiffer(leftString, rightString) {
-    Object _leftJson = new JsonDecoder().convert(leftString);
-    Object _rightJson = new JsonDecoder().convert(rightString);
-    
-    if (_leftJson is Map && _rightJson is Map) {
+  JsonDiffer(
+    String leftString,
+    String rightString, {
+    bool debug = false,
```
Comment (srawlins) [inline lib/src/json_differ.dart]: This new parameter should be documented.

File: lib/src/json_differ.dart [hunk 4]
```diff
@@ -61,68 +82,79 @@ class JsonDiffer {
...
-  DiffNode _diffObjects(Map<String,Object> left, Map<String,Object> right) {
-    DiffNode node = new DiffNode();
+  DiffNode _diffObjects(
+      Map<String, Object> left, Map<String, Object> right, String path) {
+    final node = DiffNode();
     _keepMetadata(node, left, right);
     left.forEach((String key, Object leftValue) {
-      if (ignored.contains(key))
+      if (debug) {
+        print('Diffing object "$key" at path "$path"');
```
Comment (srawlins) [inline lib/src/json_differ.dart]: I don't think these debug statements provide any value to the library. It seems like these should just be added in appropriate places according to whatever bug someone is trying to debug.

File: pubspec.yaml [hunk 5]
```diff
@@ -5,9 +5,10 @@ description: Utilities for diffing two JSON objects
 homepage: https://github.com/google/dart-json_diff
 
 environment:
-  sdk: '>=2.2.0 <3.0.0'
+  sdk: '>=2.8.4 <3.0.0'
```
Comment (srawlins) [inline pubspec.yaml]: Why does this code require 2.8.4?

File: pubspec.yaml [hunk 6]
```diff
@@ -5,9 +5,10 @@ description: Utilities for diffing two JSON objects
...
 dependencies:
-  collection: ^1.14.11
+  collection: ^1.14.13
   diff_match_patch: ^0.3.0
 dev_dependencies:
-  test: ^1.6.4
+  test: ^1.15.3
```
Comment (srawlins) [inline pubspec.yaml]: Why does this code require a newer version of test?

**Review by srawlins — CHANGES_REQUESTED** [review]

File: pubspec.yaml (hunk 5 above)
Comment (Schwusch) [reply pubspec.yaml]: You're right, technically the requirements are 2.3.0 for the collection spreads

File: pubspec.yaml (hunk 6 above)
Comment (Schwusch) [reply pubspec.yaml]: That was me making a bit broad update strokes. Probably not necessary.

File: lib/src/json_differ.dart (hunk 4 above)
Comment (Schwusch) [reply lib/src/json_differ.dart]: You're right, it's basically only useful when developing this package.

File: lib/src/json_differ.dart (hunk 3 above)
Comment (Schwusch) [reply lib/src/json_differ.dart]: I'll remove it altogether.

**Review by srawlins — APPROVED** [review]
I love it. Thanks!

**Pull request merged**